### ISSUE TITLE
useWindowDimensions() and Dimensions.get("window") both return { width: 0, height: 0 } when window does not have keyboard focus

### ISSUE THREAD

**Issue by shirakaba** [microsoft/react-native-macos#2296]
### Environment

```markdown
react-native -v: 0.73.9
npm ls react-native-macos: 0.73.34 (but affects latest `main` as well)
node -v: v22.11.0
npm -v: v10.9.0
yarn --version: 1.22.19
xcodebuild -version: Xcode 16.1, Build version 16B40
```

### Steps to reproduce the bug

1. Add a call to `useWindowDimensions()` in your JS code.
2. In Xcode, set a breakpoint inside `applicationDidFinishLaunching`. This is a simple way to prevent the app launching with keyboard focus.
3. Place a breakpoint on the call to `[NSApp keyWindow]` inside the `RCTKeyWindow` method in `node_modules/react-native-macos/React/Base/RCTUtils.m`.
4. See how it returns `nil` for `[NSApp keyWindow]` when no window has keyboard focus (`keyWindow` is `nil` [when there is no window receiving keyboard events](https://developer.apple.com/documentation/appkit/nsapplication/keywindow?language=objc)):
    ![Image](https://github.com/user-attachments/assets/227b1c69-f8dc-4bb2-bf45-7e428a619b0d)
5. This leads to downstream functions falling back to dimensions of `CGSizeZero`:
    ![Image](https://github.com/user-attachments/assets/cba385d3-68f8-446b-9080-5e066e6ff02c)

### Expected Behavior

The dimensions of the "main window" should be reliably read, even when the window does not have keyboard focus.

### Actual Behavior

Dimensions are read as `{ width: 0, height: 0 }` despite the main window's size being non-zero.

### Additional context

I've fixed it in my apps with the following pnpm patch:

```diff
diff --git a/React/Base/RCTUtils.m b/React/Base/RCTUtils.m
index 36916fc85429f40771aec650f8b1fdf422708aca..aab38f7b9d8f74cccd3094f393643c65b86bbdf0 100644
--- a/React/Base/RCTUtils.m
+++ b/React/Base/RCTUtils.m
@@ -620,7 +620,30 @@ BOOL RCTRunningInAppExtension(void)
 
   return nil;
 #else // [macOS
-  return [NSApp keyWindow];
+  // Formerly, we were returning `[NSApp keyWindow]` from here. However, it is
+  // not completely dependable:
+  //
+  // 1) keyWindow is `nil` "when there is no window receiving keyboard events":
+  //    https://developer.apple.com/documentation/appkit/nsapplication/keywindow?language=objc
+  // 2) Upon a full reload (e.g. pressing the 'R' button in the CLI), there are
+  //    briefly two windows (as it cycles between the two), and the keyWindow
+  //    value actually briefly points to the older of the two.
+  //
+  // So in this patch, we instead just pick the first window in the array.
+  // - In single-window apps, when there is just one window in the array, that
+  //   will be the "main" window.
+  // - During hot reloads, when single-window apps can briefly have two windows
+  //   in the array, the first window will be the "old" main window while the
+  //   second window will be the "new" main window. We take dimensions from the
+  //   old main window rather than the new one because the old one has already
+  //   had a sensible size set, while the new one is still at zero size.
+  //
+  //
+  // Yes, this logic will fall apart if you have a genuine multi-window app, but
+  // react-native-macos doesn't have first-class support for multi-window apps
+  // in the first place. And for now in our app, this patch solves more problems
+  // than it creates.
+  return [NSApplication.sharedApplication.windows firstObject];
 #endif // macOS]
 }
```

Evidence of fix:

1. When `[NSApp keyWindow]` is `nil`, we pick from the list of available windows instead. In the screenshot, we pick the `lastObject` (the new window during a hot reload), but later I changed the patch to take `firstObject` because the old window had the desired dimensions while the new one genuinely was `CGSizeZero`.
    ![Image](https://github.com/user-attachments/assets/500742c4-19db-46e4-a762-37c73f9bd971)
2. Downstream functions get a proper window size:
    ![Image](https://github.com/user-attachments/assets/950f1d8e-876b-4b8b-b54c-c7716fe3869b)

**Comment by shirakaba** [comment]
That's the main reason I didn't open a PR, yeah, as it doesn't handle multi-window experiences.

I think basically the original `react-native` JS APIs, assuming a single-window world, are too limited and we do need a macOS-specific one where you can specify which root view you're talking about (maybe you can set some kind of key on the root view from the native side). It would be just as necessary for multi-window Windows apps.

I think if we did add my patch as a fallback, it would basically trade some problems for others. I guess the ideal approach would be:

1. The starter template sets a default key on the root view.
2. Instead of taking `[NSApp keyWindow]`, we cycle over all windows in the application, filtering by the ones that have an RCTRootView with the default key.
3. If there are multiple matches, assume it's a hot reload and choose the earlier match in the array (as the later match in the array, although being the newer window, will probably not have had its size set).
4. Additionally provide more desktop-appropriate APIs like `useNamedWindowDimensions("main")` and `Dimensions.getNamedWindow("main")`.
